Core/src/Logger/AppEngineFlexHandlerFactory.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+namespace Google\Cloud\Core\Logger;
+
+use Exception;
+use Monolog\Logger;
+
+/**
+ * Factory to build out an AppEngineFlexHandler for the installed version of Monolog.
+ */
+class AppEngineFlexHandlerFactory
+{
+    /**
+     * Builds out an AppEngineFlexHandler for the installed version of Monolog.
+     *
+     * @param int $level [optional] The minimum logging level at which this
+     *        handler will be triggered.
+     * @param Boolean $bubble [optional] Whether the messages that are handled
+     *        can bubble up the stack or not.
+     * @param int|null $filePermission [optional] Optional file permissions
+     *        (default (0640) are only for owner read/write).
+     * @param Boolean $useLocking [optional] Try to lock log file before doing
+     *        any writes.
+     * @param resource|string|null $stream [optional]
+     *
+     * @throws Exception
+     *
+     * @return AppEngineFlexHandler|AppEngineFlexHandlerV2
+     */
+    public static function build(
+        $level = Logger::INFO,
+        $bubble = true,
+        $filePermission = 0640,
+        $useLocking = false,
+        $stream = null
+    ) {
+        $version = defined('Monolog\Logger::API') ? Logger::API : 1;
+
+        switch ($version) {
+            case 1:
+                return new AppEngineFlexHandler($level, $bubble, $filePermission, $useLocking, $stream);
+            case 2:
+                return new AppEngineFlexHandlerV2($level, $bubble, $filePermission, $useLocking, $stream);
+            default:
+                throw new Exception('Version not supported');
```

Comment:
@jdpedrie We could probably increase the overall code coverage by adding something an additional parameter to specify a forced Monolog Version and adding a unit test for an invalid version.

Another way could be to rewrite it like this

```php
$version = defined('Monolog\Logger::API') ? Logger::API : 1;

if ($version === 1) {
    return new AppEngineFlexHandler($level, $bubble, $filePermission, $useLocking, $stream);
}

return new AppEngineFlexHandlerV2($level, $bubble, $filePermission, $useLocking, $stream);
```

As long as the Monolog interfaces don't change in a 3.x release, we should be safe. What do you think?

Reply:
I'm satisfied with the coverage we have now, regardless of the failing status check. We can override that to merge. Both handlers are covered by different build matrices, it doesn't make sense to me to go further down the road of, as you mention, forcing our code into contortions for the benefit of the coverage tool.

A theoretical monolog v3 would already not be installable based on the composer requires, so we can cross that bridge another day when it becomes a reality. :smile:

ccing @dwsupplee in case he has any opposing thoughts.

Reply:
The current approach LGTM :)

Reply:
> A theoretical monolog v3 would already not be installable based on the composer requires, so we can cross that bridge another day when it becomes a reality. 😄

@dwsupplee Monolog v3 is a reality, how would you folks would prefer to handle it? Adding yet another entry in this switch case is relatively easy but is this the way you want to go?

Reply:
I already have a local branch to add support, but I wasn't even able to pull in monolog v3 because there are some external dependencies that need to be updated or released first, e.g. https://github.com/phpDocumentor/Reflection/pull/247

Reply:
@dwsupplee @lcobucci I started a PR at https://github.com/googleapis/google-cloud-php/pull/5334 🤞